agent/consul/state/config_entry_intention.go

```diff
@@ -101,8 +101,17 @@ func (s *ServiceIntentionSourceIndex) FromObject(obj interface{}) (bool, [][]byt
 
 	vals := make([][]byte, 0, len(ixnEntry.Sources))
 	for _, src := range ixnEntry.Sources {
-		sn := src.SourceServiceName()
-		vals = append(vals, []byte(sn.String()+"\x00"))
+		peer := src.Peer
+		if peer == "" {
+			peer = structs.LocalPeerKeyword
+		}
```

Comment:
we could simply choose not to write this index if there is a peer, but I thought we might eventually need to support it